### ISSUE TITLE
Feature Idea: `salve_ipc.proper`

### ISSUE THREAD

**Issue by Moosems** [salve-org/salve#32]
Feature report:

Many of the functions that the server uses could be far more optimized but I don't want to mess with the main module because in the future I may rewrite it in Mojo (for fun and because I like the language, not for the speed claims) which is similar to Python and keeping a Python implementation wouldn't hurt. More than just that though, I really like the state of it and I'm proud of the work I've done and the progress I've made making the tools all on my own. That being said, speed is an option I want available to users. I therefore think that #16, #28 and this can all be pushed into one overarching plan to have a submodule that utilizes the proper tools for the job.

**Comment by Moosems** [comment]
To summarize #16 and #28 (will be closed in favor of this issue):
 - Support tree sitter highlighting but require the user to provide the highlighter and simply make a nicer interface
 - Support LSP's with sansio_lsp_client borrowing from @Akuli's (soft ping) Porcupine plugin and modifying for ease of use by user. Again, the user will provide the file path for the LSP and this will simply be a nicer interface.
 
 What this will *not* be:
  - A repository of LSP and tree sitter files/binaries that centralizes them for convenience. The convenience this provides will be with regard to interfacing these tools, but some assembly is required. 
  
With regards to tree sitter, the library `py-tree-sitter`may or may not be used. @Akuli (slightly harder ping) has stated apprehensions regarding the library to me and I value his insights so only time will tell how I actually implement this. Also, this will be the main cherry on the v1.0.0 cake so do not be surprised if this takes a long time to actually have implemented.

**Comment by Moosems** [comment]
To any future observers, the milestone was removed from all three issues and then added back to this one to prevent it saying it has been 66% finished when in reality I haven't even begun.

**Comment by Akuli** [comment]
> speed is an option I want available

You don't need a fancy new tool like Mojo to get more speed. There are many existing tools that speed up Python a lot. For example: cython, numba, mypyc.

Also, you don't need to speed up all your code. Measure what's slow and work mostly on the bottlenecks. (They are usually not where you expect them to be.)

**Comment by Moosems** [comment]
I'm not saying I want Mojo for speed, I just like the language in that regard. The tree-sitter/LSP support is for speed. I'm fine with the speed of the current parts of Salve but I imagine that it wouldn't be ideal in some situations and I also want to try this challenge, its just proving to be significantly harder than I'd've liked (closer to panic zone than challenge zone).

**Comment by Moosems** [comment]
IDEA:

[pygls](https://github.com/openlawlibrary/pygls) is a protocol that makes implementing your own Language Server incredibly easy. `salve_ipc.proper` plans to offer an LSP and Tree-Sitter alternative with a simple API similar to what is already offered. Currently Salve offers the following commands:
 - autocompletions
 - replacements
 - syntax highlighting
 - editorconfig info
 - definitions

For some of these commands most of the inputs are already covered by the Language Server Protocol, some not. Salve IPC will already have two different kinds of servers running on the backend (the front end will remain the same no matter what info you are wanting, you will just change a command string from one to another and modify your inputs accordingly) for tree sitter and the client protocol for LSP. The tree-sitter server will likely be based on the existing communication protocol existing and I propose we move, when the tree sitter and LSP sections are stable, the default autocompletions, editorconfig info, and definitions code to that server and then move the replacements and autocompletions functions to a special language agnostic Language Server that acts as a default. After that, we drop the proper and in the backend we are left with two protocols: the custom multiprocessing communications and the LSP communications. Both running from the same frontend that wrangles the backends into shape to give the user an amazing experience.

Detailing this out a bit deeper with regards to the default syntax highlighter and definitions currently in use: these are fall backs when a proper tree sitter syntax highlighter or LSP is not provided by the user. The burden of using nice tools like good LSP's and tree-sitter highlighters falls to them to provide the command line tools that salve interfaces with. Salve is not a repository/hive/hub of tree-sitter highlighters or LSP's, it is a (much) nicer interface to the messy problems presented to people out there right now. It does, however, provide its own defaults if they cannot manage to provide their own solutions.
```python
from salve_ipc import IPC

def main() -> None:
    context = IPC()
    
    # If none is provided we always default to the basic one of file type "all". 
    # NOTE: it will not allow two servers for the same file type and won't open new ones when
    # the same command is provided
    context.add_lsp("ruff server", "python")
    
    # NOTE: .set_root_folder() and .set_open_file() will become mandatory regardless of if the user
    # specifies a custom/alternative LSP
    # NOTE: if the root folder is changed most state of the context will be lost
    # This means all files are closed and all LSP's closed (preventing UB)
    context.set_root_folder("/path/to/project")
    
    # NOTE: LSP reads the on its own but the current multiprocessing implementation requires a 
    # string so code snippets outside of files can be used. 
    # NOTE: Internally context will handle the LSP file opened, file edited, file 
    # saved parts
    # The first file added will be set as the current file open by default until a second one 
    # is specified manually
    # If we try to add aa file without setting a root folder we set the root folder to be at
    # the same folder of the file added
    context.update_file("some/local.file", open("some/local.file", "r+").read())
    
    # To maintain LSP compatibility we need to have an open file set before we can request stuff
    # NOTE: setting the open file to one not added will cause all open files to close and all 
    # LSP's to be closed before returning an Exception (I would've liked there to be a more 
    # graceful alternative but I want to prevent UB and this is the only way to safely warn the 
    #  user (I technically could use a logger but I'm not going to force the user to use 
    # them)). No open files? We just return the Exception!
    context.set_open_file("some/local.file")
    
    # The two highlighter options are str ("default") | py_tree_sitter.Parser |
    # py_tree_sitter.Language
    # If the user requests a language not yet added and doesn't choose the default highlighter
    # theres two routes of errors that could be tried: fail silently, return a response 
    # labelled cancelled, or throw an error. In the current implementation the server returns a 
    # cancelled response but the server could forge one on its own. Failing silently is not
    # something I want to do though. Between throwing an error or the fake cancellation,
    # I'll choose the fake cancellation
    context.request(
        "highlight",
        "some_obscure_language", # Can also be set to "file autodetect" to 
        # use whatever language pygments suggests
        highlighter="default",
        text_range=(1, 30)
    )
    
    # Type 1 of 2:
    # Chooses to request without a specified language (which bypasses the LSP)
    # Because of this the user is required to provide fallback builtin keywords
    context.request("autocompletions", language=None, builtin_keywords=["from", "import", "print"])
    
    # Type 2 of 2:
    # Chooses to request a specified language which has provided an LSP
    # The fake cancellation deal is the same as before if the language is not a previously given
    # LSP. In the case that the users requests the same thing twice we allow both to proceed to the
    # server and the server will probably cancel it but if the LSP chooses to give both responses
    # we let the user deal with all the responses
    context.request("autocompletions", language="python")
    
    # We close files to ensure the LSP doesn't waste memory holding the file
    # Double closing does nothing as the server will just return early if its not in the list of
    # files
    context.close_file("some/local.file")
    
    # We can manually close LSP's
    # Double closing LSP's also does nothing and if its not there it also does nothing like
    # with files
    context.close_lsp("python")
    
    # Will kill all LSP's and close all files. This is registered with atexit in case the user
    # doesn't do it manually but its always better that the user does it anyway
    # Running the kill ipc a second time will just return and not throw an error as theres nothing
    # to kill
    context.kill_ipc()
    
if __name__ == "__main__":
    main()
```

This isn't fully mapped out yet but I think the vision is definitely there and its a promising idea. It would certainly make for a v1.0.0 release. Remember that without all these comments it also fits into this small example:
```python
from salve_ipc import IPC

def main() -> None:
    context = IPC()
    context.add_lsp("ruff server", "python")
    context.set_root_folder("/path/to/project")
    context.update_file("some/local.file", open("some/local.file", "r+").read())
    context.set_open_file("some/local.file")
    context.request(
        "highlight",
        "some_obscure_language", # Can also be set to "file autodetect" to
        # use whatever language pygments suggests
        highlighter="default",
        text_range=(1, 30)
    )
    context.request("autocompletions", language=None, builtin_keywords=["from", "import", "print"])
    context.request("autocompletions", language="python")
    context.close_file("some/local.file")
    context.close_lsp("python")
    context.kill_ipc()

if __name__ == "__main__":
    main()
```

Which in can sometimes be shortened to just:
```python
from salve_ipc import IPC

def main() -> None:
    context = IPC()
    context.set_root_folder("/path/to/project")
    context.update_file("some/local.file", open("some/local.file", "r+").read())
    context.request("autocompletions", language=None, builtin_keywords=["from", "import", "print"])
    context.kill_ipc()

if __name__ == "__main__":
    main()
```

Not definitive on this idea but it seems nicer the longer I think about it :D. I get to keep all my custom-made tools and also introduce nicer tools that people generally want all with a very nice interface easily embeddable into editors!

**Comment by Moosems** [comment]
Alternatively, we could keep the custom functions in a the server and keep the new API for the Client side of LSP handling

**Comment by Moosems** [comment]
Crazy idea, but instead of trying to cram all of this into one tool, I can make four separate but similar tools. The three tools would be as follows:
 - salve (the basic tools but not much customization needed. It "Just Works" and makes everything super simple)
 - simple-tree-sitter (a nice wrapper around tree-sitter that makes the job easier (combined with the salve-dependency-hub)
 - py-simple-lsp (a good lsp client that makes the job easy but still requires some work from the user)
 - py-PEG (a good PEG library thats easy to use and has some builtin grammars)
 
 Then, I can wrap this all under one organization named Salve which mainly prioritizes the development of the Salve package but also maintains the other three.

@leycec, @Akuli, you both have more experience in this domain than I, what do you think of something like this?

**Comment by Moosems** [comment]
I have created and reserved the organization name `salve-org` as the name `Salve` is taken by another user. Won't transfer repository until a discussion on this occurs.

**Comment by Moosems** [comment]
Going to go ahead with this as I'm a little impatient and want to keep progress going

**Comment by Moosems** [comment]
#73 is a part of this.

**Comment by Moosems** [comment]
Going to make this part of a salve-org discussion on future plans and remove the stuff albero now covers from the v1.0.0 branch. #74 Will be part of v1.0.0

**Comment by Moosems** [comment]
Closing in favor of [Salve Organization Discussion #1](https://github.com/orgs/salve-org/discussions/1)
